### ISSUE TITLE
Support apollo-client v2

### ISSUE THREAD

**Issue by AdamYee** [jaydenseric/apollo-upload-client#33]
Hi Jayden!

I think I may have read you're already working on this (i.e. creating a custom Link)? Do you need any help with this? https://github.com/apollographql/apollo-link/blob/master/docs/implementation.md

Adam

**Comment by jaydenseric** [comment]
[`apollo-upload-fetch`](https://www.npmjs.com/package/apollo-fetch-upload) is good to go; it supersedes `apollo-upload-client` for `apollo-client` v2. I have been using the new link based network interface for a while now, you can see how it works together [in `apollo-upload-examples`](https://github.com/jaydenseric/apollo-upload-examples/blob/master/app/helpers/with-data.js#L23).

I you would like to help out, the examples could do with an update to `apollo-client` v2. The [upgrade guide](https://github.com/apollographql/apollo-client/blob/master/Upgrade.md) has been improved since I got stuck on some `getInitialState` API changes. I'm waiting for a new release of `react-apollo` that has an `apollo-client` v2 dependency before upgrading my own app, to avoid multiple versions of `apollo-client` getting bundled.

**Comment by jaydenseric** [comment]
An update: Apollo Link now uses the native fetch implementation instead of `apollo-fetch`. Once I am sure of the situation I'll likely update `apollo-upload-client` for Apollo Client v2 sometime soon (within the next week hopefully) and deprecate `apollo-fetch-upload`. Only the setup for Apollo Client will change; everything else should work the same as before.

**Comment by jaydenseric** [comment]
I have been working on an upload link for Apollo Client v2 (keep an eye on the [`next` branch](https://github.com/jaydenseric/apollo-upload-client/tree/next)) but it has been trickier than I anticipated.

- [`apollo-link-http`](https://github.com/apollographql/apollo-link/tree/master/packages/apollo-link-http) [encodes and sets the request body itself](https://github.com/apollographql/apollo-link/blob/apollo-link-http%401.0.0/packages/apollo-link-http/src/httpLink.ts#L128) and [forcibly sets the content-type to JSON](https://github.com/apollographql/apollo-link/blob/apollo-link-http%401.0.0/packages/apollo-link-http/src/httpLink.ts#L148). This means it will have to be entirely replaced by a new terminating link as we set the body to `FormData` and use a different content-type. This is not a huge surprise, but I would like to match the API as close as possible and there are some parts I am not familiar with (e.g. [cancelable requests](https://github.com/apollographql/apollo-link/blob/apollo-link-http%401.0.0/packages/apollo-link-http/src/httpLink.ts#L183)).
- [`apollo-link-batch-http`](https://github.com/apollographql/apollo-link/tree/master/packages/apollo-link-batch-http) still uses `apollo-fetch`, and it could be weeks before it is updated to a final API. I am not sure the right way to support batching with an upload link. Will we need to publish and maintain a separate batching upload link?

In the meantime, [`apollo-fetch-upload`](https://www.npmjs.com/package/apollo-fetch-upload) works fine with Apollo Client v2, and if you use batching you will still need to use `apollo-fetch` for a while anyway. I am focusing on [a new generation spec for multipart GraphQL requests](https://github.com/jaydenseric/graphql-multipart-request-spec/pull/1) along with a new version of `apollo-upload-server` for now.

**Comment by leethree** [comment]
> apollo-fetch-upload works fine with Apollo Client v2

Hi @jaydenseric , could you provide an example for that? I can't seem to make it work using Apollo v2 without batching.

**Comment by AdamYee** [comment]
I think it would be fine to have a separate publish for an `apollo-link-batch-upload`.

**Comment by jaydenseric** [comment]
I've got a non-batching upload terminating link working [in the `next` branch](https://github.com/jaydenseric/apollo-upload-client/blob/next/src/index.mjs).

With a bit more polish and testing we'll be able to publish a beta.

The API diverges slightly from `apollo-link-http`:

- I can't see any value in exporting a class like [here](https://github.com/apollographql/apollo-link/blob/apollo-link%401.0.0/packages/apollo-link-http/src/httpLink.ts#L191).
- There are several error checks and messages I don't think are worth the complexity and increased bundle size. For example, [checking if Apollo Fetch is accidentally being used](https://github.com/apollographql/apollo-link/blob/apollo-link%401.0.0/packages/apollo-link-http/src/httpLink.ts#L44) and [instructions for polyfilling global `fetch`](https://github.com/apollographql/apollo-link/blob/apollo-link%401.0.0/packages/apollo-link-http/src/httpLink.ts#L60).
- There is a confusing [`ResponseError` type](https://github.com/apollographql/apollo-link/blob/apollo-link%401.0.0/packages/apollo-link-http/src/httpLink.ts#L11) used in a few places that does not seem to be a documented part of the Apollo Link API, i.e. important to [`apollo-link-error`](http://apollo-link-docs.netlify.com/docs/link/links/error.html). I chose not to adopt it and the way several errors are thrown.
- There is a fair bit of [complexity in `apollo-link-http`](https://github.com/apollographql/apollo-link/blob/apollo-link%401.0.0/packages/apollo-link-http/src/httpLink.ts#L153) around the multiple ways to set fetch options such as `credentials` and `headers`. IMO all fetch options should be set with the `fetchOptions` link initialization option and context property only.
- I don't know much about the new [`AbortController`](https://github.com/apollographql/apollo-link/blob/apollo-link%401.0.0/packages/apollo-link-http/src/httpLink.ts#L78) or how it can be used with Apollo links, so I chose not to implement it.

**Comment by jaydenseric** [comment]
@leethree I don't have an example handy sorry, I have never not used batching. Perhaps it is not easy or possible, in which case it might be best to use batching until this update is published.

This is something like my 4th implementation of uploads for Apollo Client 😓

**Comment by VladShcherbin** [comment]
@jaydenseric hey, since new apollo packages are growing like mushrooms after rain, what's the best way for do file uploads now? What stable packages combo would you suggest and with or without batching?

p.s. sorry you have to re-implement this every time apollo decides to change internals, it really sucks.

**Comment by jaydenseric** [comment]
Just published [v6.0.0-beta.1](https://github.com/jaydenseric/apollo-upload-client/releases/tag/v6.0.0-beta.1) 🚀

It does not require any changes to your codebase other than a new Apollo Client setup. I'll update [apollo-upload-examples](https://github.com/jaydenseric/apollo-upload-examples) shortly.

Unfortunately batching is not available in this release; hopefully we can support it soon in an update. I'm waiting on a solid batching link from Apollo to reference. If you desperately need batching, you can continue to use [apollo-fetch-upload](https://github.com/apollographql/apollo-fetch/tree/master/packages/apollo-fetch-upload#setup) with `apollo-link-batch-http@^1`.

**Comment by jaydenseric** [comment]
Updated apollo-upload-examples: [Here is the new Apollo Client setup](https://github.com/jaydenseric/apollo-upload-examples/blob/master/app/providers/with-data.js#L24).

Closing this now we support the final Apollo Client v2 API. I will raise a new issue to add back batching support. Please everyone test out the beta and raise any issues so we can rush it into production 🙌

**Comment by Fi1osof** [comment]
> There is a confusing ResponseError type used in a few places that does not seem to be a documented part of the Apollo Link API, i.e. important to apollo-link-error. I chose not to adopt it and the way several errors are thrown.

Why not return raw apollo error? I can not catch GraphQL errors. In raw error object exists:
operation: The Operation that errored
**response**: The result returned from lower down in the link chain
graphQLErrors: An array of errors from the GraphQL endpoint
networkError: Any error during the link execution or server response, that wasn’t delivered as part of the **errors** field in the GraphQL result

In apollo-upload-client just operation, graphQLErrors (null), networkError. response required.
